netplugin/nameserver/nameserver.go

```diff
@@ -0,0 +1,705 @@
+/***
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package nameserver
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/Sirupsen/logrus"
+	"github.com/contiv/netplugin/core"
+	"github.com/contiv/netplugin/netmaster/mastercfg"
+	"github.com/miekg/dns"
+	cmap "github.com/streamrail/concurrent-map"
+	"hash/fnv"
+	"net"
+	"strings"
+	"sync"
+)
+
+// decorator log
+var dnsLog *logrus.Entry
+
+type tenantBucket struct {
+	sync.RWMutex
+	tenantTables map[string]*dnsTables
+}
+
+const commonK8sTenant = "common"
+
+// NetpluginNameServer config
+type NetpluginNameServer struct {
+	svcKeyPath  string
+	epKeyPath   string
+	epChan      chan core.WatchState
+	epErrChan   chan error
+	svcChan     chan core.WatchState
+	svcErrChan  chan error
+	stateDriver core.StateDriver
+	bucketSize  uint
+	buckets     []tenantBucket
+	commonSvc   cmap.ConcurrentMap // for non-multi tenant LB service
+	stats       struct {
+		sync.RWMutex
+		tenantStats map[string]map[string]uint64
+	}
+}
+
+// DNS name record, ipv4 & ipv6 address
+type nameRecord struct {
+	v4Record net.IP
+	v6Record net.IP
+}
+
+// dns records per tenant
+type dnsTables struct {
+	svcTbl      map[string]nameRecord      // LB service records
+	endpointTbl map[string]nameRecord      // endpoint-id records
+	epgTbl      map[string]map[string]bool // endpoint group records
+	nameTbl     map[string]map[string]bool // container-name records
+}
+
+func lookUpServiceV4Record(record nameRecord, name string) ([]dns.RR, int) {
+	rr := []dns.RR{}
+
+	if record.v4Record.IsUnspecified() != true {
+		r := func() dns.RR {
+			r := new(dns.A)
+			r.A = record.v4Record
+			r.Hdr = dns.RR_Header{Name: name + ".", Rrtype: dns.TypeA,
+				Class: dns.ClassINET, Ttl: 64}
+			return r
+		}()
+		rr = append(rr, r)
+	}
+
+	return rr, len(rr)
+}
+
+func (dt *dnsTables) lookUpEndPointV4Record(nameList map[string]bool, name string) ([]dns.RR, int) {
+	rr := []dns.RR{}
+	for i := range nameList {
+		if record, ok := dt.endpointTbl[i]; ok {
+			if record.v4Record.IsUnspecified() != true {
+				r := func() dns.RR {
```

Comment:
dns.RR is an interface. As long as type dns.A implements the dns.RR interface(which it does), you can directly assign it or use it in place of RR. There is no need of type conversion here.